### PR TITLE
Split sharding table contract

### PR DESCRIPTION
**Pull request by NZT48**
This PR splits Sharding Table contract into three different contracts:
1. ShardingTable.sol - contract that contains logic
2. ShardingTableStorage.sol - contract that handles data
3. IShardingTableStructs.sol - interface that contains Node and NodeInfo structs that are shared between sharding table and sharding table contracts

To do:
- [x] Implement sharding table storage
- [x] Implement interface for sharing structs between files
- [x] Implement tests for sharding table

### CODE REVIEW

**Review by u-hubar — CHANGES_REQUESTED** [review]
Few changes needed, but overall logic is good.

I also propose sticking to these rules: https://docs.soliditylang.org/en/latest/style-guide.html

File: contracts/storage/ShardingTableStorage.sol [hunk 1]
```diff
@@ -0,0 +1,252 @@
...
+
+contract ShardingTableStorage is IShardingTableStructs {
+
+    event NodeObjCreated(uint96 indexed identityId, bytes nodeId, uint96 ask, uint96 stake);
+    event NodeRemoved(uint96 indexed identityId, bytes nodeId);
+    event NodeRemovedByHubOwner(bytes nodeId);
+
+    Hub public hub;
+
+    bytes private emptyPointer;
```
Comment (u-hubar) [inline contracts/storage/ShardingTableStorage.sol:17]: We can make it constant I guess

File: contracts/storage/ShardingTableStorage.sol [hunk 2]
```diff
@@ -0,0 +1,252 @@
...
+            emptyPointer
+        );
+
+        nodes[nodeId] = newNode;
+        nodeIdsSha256[nodeId] = nodeIdSha256;
+
+        emit NodeObjCreated(identityId, nodeId, profileStorage.getAsk(identityId), profileStorage.getStake(identityId));
+    }
+
+    function getNode(bytes memory nodeId)
```
[the remark this answers is not in the document]
Comment (NZT48) [reply contracts/storage/ShardingTableStorage.sol:150]: Solidity cannot generate proper built-in getters for complex structures, more info here: https://docs.soliditylang.org/en/v0.8.13/contracts.html#getter-functions

File: contracts/storage/ShardingTableStorage.sol [hunk 3]
```diff
@@ -0,0 +1,252 @@
...
+        Node memory currentNode = nodes[firstNodeId];
+        for (uint16 i = 0; i < nodesNumber; i++) {
+            nodesPage[i] = currentNode;
+            currentNode = nodes[currentNode.nextNodeId];
+        }
+
+        return nodesPage;
+    }
+
+    function getNodeIdsSha256(bytes memory nodeId)
```
[the remark this answers is not in the document]
Comment (NZT48) [reply contracts/storage/ShardingTableStorage.sol]: same as for previous comment

File: contracts/storage/ShardingTableStorage.sol [hunk 4]
```diff
@@ -0,0 +1,252 @@
...
+        require (
+            msg.sender == hub.owner(),
+            "Function can only be called by hub owner!"
+        );
+        _;
+    }
+
+    function pushBack(uint96 identityId)
+        public
+        onlyContracts
```
[the remark this answers is not in the document]
Comment (NZT48) [reply contracts/storage/ShardingTableStorage.sol:43]: As we discussed, we don't want to lock so sharding table contract can access sharding table storage, we want to have possibility that other DKG contracts can access it in future if it is needed.

File: contracts/storage/ShardingTableStorage.sol [hunk 5]
```diff
@@ -0,0 +1,252 @@
...
+        hub = Hub(hubAddress);
+
+        emptyPointer = "";
+        nodeIdsSha256[emptyPointer] = sha256(emptyPointer);
+        head = emptyPointer;
+        tail = emptyPointer;
+        nodesCount = 0;
+    }
+
+    modifier onlyContracts(){
```
[the remark this answers is not in the document]
Comment (NZT48) [reply contracts/storage/ShardingTableStorage.sol:35]: As we discussed, we don't want to lock so sharding table contract can access sharding table storage, we want to have possibility that other DKG contracts can access it in future if it is needed.

**Review by branarakic — APPROVED** [review]

File: contracts/storage/ShardingTableStorage.sol [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+    }
+
+    function setTail(bytes memory nodeId)
+        public
+        onlyContracts
+    {
+        tail = nodeId;
+    }
+
+    function getNode(bytes memory nodeId)
```
Comment (u-hubar) [inline contracts/storage/ShardingTableStorage.sol:68]: Can be removed?

File: test/shardingTable.test.js [hunk 7]
```diff
@@ -1,42 +1,52 @@
...
 
 const ERC20Token = artifacts.require('ERC20Token');
 const Hub = artifacts.require('Hub');
 const Profile = artifacts.require('Profile');
 const ShardingTable = artifacts.require('ShardingTable');
+const ShardingTableStorage = artifacts.require('ShardingTableStorage');
+const IdentityStorage = artifacts.require('IdentityStorage');
+const ProfileStorage = artifacts.require('ProfileStorage');
+const ParametersStorage = artifacts.require('ParametersStorage');
+const bytes32 = require('bytes32');
```
Comment (u-hubar) [inline test/shardingTable.test.js:12]: Do we need this?

File: package.json [hunk 8]
```diff
@@ -27,7 +27,8 @@
     "eslint": "^8.23.0",
     "ganache": "^7.4.4",
     "solhint": "^3.3.7",
-    "truffle": "^5.6.0"
+    "truffle": "^5.6.0",
+    "bytes32": "^0.0.3"
```
Comment (u-hubar) [inline package.json:31]: Do we need this?

File: contracts/ShardingTable.sol [hunk 9]
```diff
@@ -103,108 +92,142 @@ contract ShardingTable {
...
+
+        shardingTableStorage.decrementNodesCount();
 
-        nodesCount -= 1;
-    
         emit NodeRemovedByHubOwner(nodeId);
     }
 
+    function equalIdHashes(bytes memory firstId, bytes memory secondId)
+        public
```
Comment (u-hubar) [inline contracts/ShardingTable.sol:217]:
```suggestion
        private
```

File: contracts/ShardingTable.sol [hunk 10]
```diff
@@ -103,108 +92,142 @@ contract ShardingTable {
...
+        shardingTableStorage.removeNode(nodeId);
+
+        shardingTableStorage.decrementNodesCount();
 
-        nodesCount -= 1;
-    
         emit NodeRemovedByHubOwner(nodeId);
     }
 
+    function equalIdHashes(bytes memory firstId, bytes memory secondId)
```
Comment (u-hubar) [inline contracts/ShardingTable.sol:216]:
```suggestion
    function _equalIdHashes(bytes memory firstId, bytes memory secondId)
```

**Review by u-hubar — CHANGES_REQUESTED** [review]

File: contracts/storage/ShardingTableStorage.sol [hunk 11]
```diff
@@ -0,0 +1,123 @@
...
+
+    bytes public constant emptyPointer = "";
+
+    Hub public hub;
+
+    bytes public head;
+    bytes public tail;
+    uint16 public nodesCount;
+
+    mapping(bytes => Node) nodes;
```
[the remark this answers is not in the document]
Comment (NZT48) [reply contracts/storage/ShardingTableStorage.sol:19]: Solidity cannot generate proper built-in getters for complex structures, more info here: https://docs.soliditylang.org/en/v0.8.13/contracts.html#getter-functions

File: test/shardingTable.test.js (hunk 7 above)
Comment (NZT48) [reply test/shardingTable.test.js:12]: There is still more work to be done on tests, this should help with crafting values, tests with its usage will be pushed with next commit. Don't focus too much on tests yet.

File: package.json (hunk 8 above)
Comment (NZT48) [reply package.json:31]: There is still more work to be done on tests, this should help with crafting values, tests with its usage will be pushed with next commit. Don't focus too much on tests yet.

File: contracts/storage/ShardingTableStorage.sol (hunk 11 above)
[the remark this answers is not in the document]
Comment (u-hubar) [reply contracts/storage/ShardingTableStorage.sol:19]: I believe it only applies to structures with nested mappings

File: test/shardingTable.test.js (hunk 7 above)
Comment (u-hubar) [reply test/shardingTable.test.js:12]: > There is still more work to be done on tests, this should help with crafting values, tests with its usage will be pushed with next commit. Don't focus too much on tests yet.

Np. I think that we should be consistent and use ethers.js for different kinds of type conversions etc. as we now do in other repos

File: contracts/storage/ShardingTableStorage.sol (hunk 11 above)
[the remark this answers is not in the document]
Comment (NZT48) [reply contracts/storage/ShardingTableStorage.sol:19]: Built in getters returns tuple in this format: `(uint72 identityId, bytes memory id, bytes memory prevNodeId, bytes memory nextNodeId)`, and we then need to deconstruct this tuple, crate new Node object and pack it into. We use getter for node 14 times in `ShardingTable.sol` so using default getters will significantly increase code size and decrease readability.

File: test/shardingTable.test.js (hunk 7 above)
Comment (NZT48) [reply test/shardingTable.test.js:12]: Good point

File: contracts/storage/ShardingTableStorage.sol [hunk 12]
```diff
@@ -0,0 +1,123 @@
...
+
+pragma solidity ^0.8.0;
+
+import {Hub} from '../Hub.sol';
+import {ProfileStorage} from './ProfileStorage.sol';
+import {IShardingTableStructs} from '../interface/IShardingTableStructs.sol';
+
+contract ShardingTableStorage is IShardingTableStructs {
+
+    bytes public constant emptyPointer = "";
```
[the remark this answers is not in the document]
Comment (NZT48) [reply contracts/storage/ShardingTableStorage.sol]: Applied

File: contracts/ShardingTable.sol [hunk 13]
```diff
@@ -4,49 +4,29 @@ pragma solidity^0.8.0;
...
-        bytes prevNodeId;
-        bytes nextNodeId;
-    }
-
-    struct NodeInfo {
-        bytes id;
-        uint96 ask;
-        uint96 stake;
-    }
+    bytes public constant emptyPointer = "";
```
[the remark this answers is not in the document]
Comment (NZT48) [reply contracts/ShardingTable.sol]: We are now getting it from ShardingTableStorage

**Review by u-hubar — APPROVED** [review]

**Pull request merged**